spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -767,16 +784,51 @@ private void registerComponentWithBeanFactory(String name, Object component) {
 		}
 	}
 
-	private SubscribableChannel postProcessChannelForFunction(MessageChannel outputChannel) {
-		if (integrationFlowFunctionSupport != null && integrationFlowFunctionSupport.containsFunction(Function.class)) {
-			DirectChannel actualOutputChannel = new DirectChannel();
-			integrationFlowFunctionSupport.andThenFunction(MessageChannelReactiveUtils.toPublisher(outputChannel),
-				actualOutputChannel);
-			return actualOutputChannel;
+	private SubscribableChannel postProcessOutboundChannelForFunction(MessageChannel outputChannel) {
+		if (this.integrationFlowFunctionSupport != null) {
+			Publisher publisher = MessageChannelReactiveUtils.toPublisher(outputChannel);
+			// If the app has an explicit Supplier bean defined, make that as the publisher
+			if (this.integrationFlowFunctionSupport.containsFunction(Supplier.class)) {
+				IntegrationFlowBuilder integrationFlowBuilder = IntegrationFlows.from(outputChannel).bridge();
+				publisher = integrationFlowBuilder.toReactivePublisher();
+			}
+			if (this.integrationFlowFunctionSupport.containsFunction(Function.class,
+					this.streamFunctionProperties.getDefinition())) {
+				DirectChannel actualOutputChannel = new DirectChannel();
+				if (outputChannel instanceof AbstractMessageChannel) {
+					moveChannelInterceptors((AbstractMessageChannel) outputChannel, actualOutputChannel);
+				}
+				this.integrationFlowFunctionSupport.andThenFunction(publisher, actualOutputChannel,
+						this.streamFunctionProperties.getDefinition());
+				return actualOutputChannel;
+			}
 		}
 		return (SubscribableChannel) outputChannel;
 	}
 
+	private SubscribableChannel postProcessInboundChannelForFunction(MessageChannel inputChannel) {
+		if (this.integrationFlowFunctionSupport != null &&
+				(this.integrationFlowFunctionSupport.containsFunction(Consumer.class) ||
+						this.integrationFlowFunctionSupport.containsFunction(Function.class))) {
+			DirectChannel actualInputChannel = new DirectChannel();
+			if (inputChannel instanceof AbstractMessageChannel) {
+				moveChannelInterceptors((AbstractMessageChannel) inputChannel, actualInputChannel);
+			}
+			this.integrationFlowFunctionSupport.andThenFunction(MessageChannelReactiveUtils.toPublisher(actualInputChannel),
+					inputChannel, this.streamFunctionProperties.getDefinition());
+			return actualInputChannel;
+		}
+		return (SubscribableChannel) inputChannel;
+	}
+
+	private void moveChannelInterceptors(AbstractMessageChannel existingMessageChannel,
+			AbstractMessageChannel actualMessageChannel) {
+		for (ChannelInterceptor channelInterceptor : existingMessageChannel.getChannelInterceptors()) {
+			actualMessageChannel.addInterceptor(channelInterceptor);
+			existingMessageChannel.removeInterceptor(channelInterceptor);
+		}
```

Comment:
Love it! I had raised a concern previously about apps that might put an interceptor on output. I actually tried this in one the sftp source iterations and it's not safe to assume the it will play nice with SCSt interceptors. i.e.,  the payload was already converted to byte.  IMO, if it comes up, we should advise against stream app devs putting interceptors on bound channels.